fh_immuta_utils/scripts/tag_existing_data_sources.py

```diff
@@ -77,38 +83,156 @@ def main(
+            data_source_tag_names - current_data_source_tag_names
+        )
+
+        for tag_name in new_data_source_tag_names:
+            new_tag_name_to_data_source_ids[tag_name].add(data_source["id"])
+
+        removed_data_source_tag_names = (
+            current_data_source_tag_names - data_source_tag_names
+        )
+        for tag_name in removed_data_source_tag_names:
+            removed_tag_name_and_data_source_ids.add((data_source["id"], tag_name))
+
+    logging.info("Updating new data source tags")
+    progress_iterator = tqdm(
+        new_tag_name_to_data_source_ids.items(),
+        total=len(new_tag_name_to_data_source_ids),
+    )
+    for tag_name, data_source_ids in progress_iterator:
+        progress_iterator.set_description(
+            desc=f"[Tagging Data Sources] Tag: {tag_name}, # Data Sources: {len(data_source_ids)} :"
+        )
+        if not dry_run:
+            client.update_data_source_tags_in_bulk(
+                ids=list(data_source_ids),
+                tag_data=[
+                    {
+                        "name": tag_name,
+                        "source": "curated",
+                    }
+                ],
+            )
+    logging.info("Updating removed data source tags")
+    progress_iterator = tqdm(removed_tag_name_and_data_source_ids)
+    for data_source_id, tag_name in progress_iterator:
+        progress_iterator.set_description(
+            desc=f"[Removing Data Source Tag] ID: {data_source_id}, Tag: {tag_name}"
+        )
+        if not dry_run:
+            client.delete_data_source_tag(data_source_id, tag_name)
+
+    logging.info("Determining data sources with new column tags")
+    # Compute the columns and associated column tag names we want for each data source.
+    #
+    # Since we cannot get the set of all columns of a data source without making
+    # an Immuta API call for each data source (which would be expensive), we
+    # compute the set of all relevant columns for each data source by searching
+    # for all data sources with a given relevant column. E.g. if we want to tag
+    # all columns with name `example`, we find the relevant data sources by
+    # searching for data sources with the column `example`, rather than getting
+    # the set of all columns of all data sources and then filtering the data
+    # sources.
+    #
+    # data_source_name: {column_1_name: {tag_1_name, tag_2_name, ...}, ...}
+    wanted_column_tag_names_map: Dict[str, Dict[str, Set[str]]] = defaultdict(dict)
+    for column in tqdm(
+        tagger.tag_map_datadict.keys(),
+        total=len(tagger.tag_map_datadict),
+    ):
+        tag_names = set(tagger.get_tags_for_column(column))
+        with Paginator(
+            client.get_data_source_list,
+            search_text=search_text,
+            search_schema=search_schema,
+            columns=[column],
+            size=IMMUTA_API_PAGE_SIZE,
+        ) as paginator:
+            for data_source in paginator:
```

Comment:
Could we do this work in the existing iteration over client.get_data_source_list? (line 71ish)

Reply:
I don't think so, b/c you'd miss data sources with no matching column (but should still have a data source level tag)